File: cupcake-py/python/cupcake/installer.py
"""
OPA Binary Installer for Cupcake

This module handles automatic download and verification of the OPA binary
required for policy compilation.

Version: OPA v1.7.1
"""

import hashlib
import os
import platform
import shutil
import stat
import subprocess
import sys
import urllib.request
from pathlib import Path
from typing import Optional, Tuple

# OPA Version Configuration
OPA_VERSION = "v1.7.1"
OPA_BASE_URL = f"https://github.com/open-policy-agent/opa/releases/download/{OPA_VERSION}"

# Platform-specific binary mapping with SHA256 checksums
OPA_BINARIES = {
    ("Darwin", "x86_64"): {
        "binary": "opa_darwin_amd64",
        "sha256": "51da8fa6ce4ac9b963d4babbd78714e98880b20e74f30a3f45a96334e12830bd",
        "size_mb": 67.3
    },
    ("Darwin", "arm64"): {
        "binary": "opa_darwin_arm64_static", 
        "sha256": "fe2a14b6ba7f587caeb62ef93ef62d1e713776a6e470f4e87326468a8ecfbfbd",
        "size_mb": 43.8
    },
    ("Linux", "x86_64"): {
        "binary": "opa_linux_amd64",
        "sha256": "7426bf5504049d7444f9ee9a1d47a64261842f38f5308903ef6b76ba90250b5a",
        "size_mb": 67.1
    },
    ("Linux", "aarch64"): {
        "binary": "opa_linux_arm64_static",
        "sha256": "a81af8cd767f1870e9e23b8ed0ad8f40b24e5c0a64c5768c75d5c292aaa81e54",
        "size_mb": 43.2
    },
    ("Windows", "AMD64"): {
        "binary": "opa_windows_amd64.exe",
        "sha256": "205f87d0fd1e2673c3a6f9caf9d9655290e478a93eeb3ef9f211acdaa214a9ca",
        "size_mb": 98.7
    },
}


def get_cache_dir() -> Path:
    """Get the directory for caching OPA binaries"""
    if sys.platform == "win32":
        cache_base = Path(os.environ.get("LOCALAPPDATA", Path.home() / "AppData" / "Local"))
    else:
        cache_base = Path(os.environ.get("XDG_CACHE_HOME", Path.home() / ".cache"))
    
    cache_dir = cache_base / "cupcake" / "bin"
    cache_dir.mkdir(parents=True, exist_ok=True)
    return cache_dir


def get_platform_info() -> Tuple[str, str]:
    """Detect the current platform and architecture"""
    system = platform.system()
    machine = platform.machine()
    
    # Normalize platform names
    if machine in ("x86_64", "AMD64"):
        machine = "x86_64"
    elif machine in ("arm64", "aarch64"):
        machine = "arm64" if system == "Darwin" else "aarch64"
    
    return system, machine


def verify_checksum(file_path: Path, expected_sha256: str) -> bool:
    """Verify the SHA256 checksum of a file"""
    sha256_hash = hashlib.sha256()
    with open(file_path, "rb") as f:
        for chunk in iter(lambda: f.read(4096), b""):
            sha256_hash.update(chunk)
    
    actual = sha256_hash.hexdigest()
    return actual == expected_sha256


def download_with_progress(url: str, dest_path: Path, expected_size_mb: float) -> None:
    """Download a file with progress indication"""
    print(f"Downloading OPA {OPA_VERSION} ({expected_size_mb:.1f} MB)...")
    
    with urllib.request.urlopen(url) as response:
        total_size = int(response.headers.get('Content-Length', 0))
        downloaded = 0
        
        with open(dest_path, 'wb') as f:
            while True:
                chunk = response.read(8192)
                if not chunk:
                    break
                
                f.write(chunk)
                downloaded += len(chunk)
                
                if total_size > 0:
                    percent = (downloaded / total_size) * 100
                    print(f"Progress: {percent:.1f}%", end='\r')
    
    print("\nDownload complete!")


def make_executable(path: Path) -> None:
    """Make a file executable on Unix-like systems"""
    if sys.platform != "win32":
        st = path.stat()
        path.chmod(st.st_mode | stat.S_IEXEC)


def download_opa(force: bool = False) -> Path:
    """
    Download and verify the OPA binary for the current platform
    
    Args:
        force: Force re-download even if binary exists
    
    Returns:
        Path to the OPA executable
    
    Raises:
        RuntimeError: If platform is unsupported or download fails
    """
    # Get platform info
    system, machine = get_platform_info()
    platform_key = (system, machine)
    
    if platform_key not in OPA_BINARIES:
        raise RuntimeError(
            f"Unsupported platform: {system} {machine}\n"
            f"Supported platforms: {list(OPA_BINARIES.keys())}"
        )
    
    binary_info = OPA_BINARIES[platform_key]
    binary_name = binary_info["binary"]
    expected_sha256 = binary_info["sha256"]
    size_mb = binary_info["size_mb"]
    
    # Determine local path
    cache_dir = get_cache_dir()
    local_name = f"opa-{OPA_VERSION}"
    if sys.platform == "win32":
        local_name += ".exe"
    local_path = cache_dir / local_name
    
    # Check if already downloaded and valid
    if local_path.exists() and not force:
        print(f"Verifying existing OPA binary at {local_path}...")
        if verify_checksum(local_path, expected_sha256):
            print("Checksum verified successfully!")
            return local_path
        else:
            print("Checksum verification failed, re-downloading...")
            local_path.unlink()
    
    # Download the binary
    url = f"{OPA_BASE_URL}/{binary_name}"
    temp_path = local_path.with_suffix('.tmp')
    
    try:
        download_with_progress(url, temp_path, size_mb)
        
        # Verify checksum
        print("Verifying checksum...")
        if not verify_checksum(temp_path, expected_sha256):
            raise RuntimeError(
                f"Checksum verification failed for {binary_name}\n"
                f"This could indicate a corrupted download or security issue."
            )
        
        # Move to final location
        temp_path.rename(local_path)
        make_executable(local_path)
        
        print(f"OPA {OPA_VERSION} installed successfully at {local_path}")
        return local_path
        
    except Exception as e:
        # Clean up on failure
        if temp_path.exists():
            temp_path.unlink()
        raise RuntimeError(f"Failed to download OPA: {e}")


def find_opa() -> Optional[Path]:
    """
    Find OPA binary in order of preference:
    1. CUPCAKE_OPA_PATH environment variable
    2. Cached download
    3. System PATH
    
    Returns:
        Path to OPA or None if not found
    """
    # Check environment variable
    env_path = os.environ.get("CUPCAKE_OPA_PATH")
    if env_path:
        path = Path(env_path)
        if path.exists():
            return path
    
    # Check cache
    cache_dir = get_cache_dir()
    local_name = f"opa-{OPA_VERSION}"
    if sys.platform == "win32":
        local_name += ".exe"
    cached_path = cache_dir / local_name
    if cached_path.exists():
        return cached_path
    
    # Check system PATH
    opa_cmd = "opa.exe" if sys.platform == "win32" else "opa"
    opa_path = shutil.which(opa_cmd)
    if opa_path:
        # Verify it's the right version
        try:
            result = subprocess.run(
                [opa_path, "version"],
                capture_output=True,
                text=True,
                timeout=5
            )
            if OPA_VERSION.lstrip('v') in result.stdout:
                return Path(opa_path)
        except:
            pass
    
    return None


def ensure_opa_installed() -> Path:
    """
    Ensure OPA is installed and return its path
    
    This is the main entry point for the installer.
    It will use an existing OPA if found, or download it if needed.
    
    Returns:
        Path to the OPA executable
    
    Raises:
        RuntimeError: If OPA cannot be installed
    """
    # Try to find existing OPA
    opa_path = find_opa()
    if opa_path:
        return opa_path
    
    # Download OPA
    return download_opa()


def get_opa_command() -> str:
    """
    Get the OPA command for subprocess calls
    
    Returns:
        String path to OPA executable
    """
    opa_path = ensure_opa_installed()
    return str(opa_path)


# Module-level convenience
if __name__ == "__main__":
    # Allow running as a script for manual installation
    import argparse
    
    parser = argparse.ArgumentParser(description="Install OPA for Cupcake")
    parser.add_argument("--force", action="store_true", help="Force re-download")
    parser.add_argument("--version", action="store_true", help="Show version")
    
    args = parser.parse_args()
    
    if args.version:
        print(f"OPA version: {OPA_VERSION}")
    else:
        try:
            path = download_opa(force=args.force)
            print(f"OPA is available at: {path}")
        except Exception as e:
            print(f"Error: {e}", file=sys.stderr)
            sys.exit(1)

File: cupcake-ts/installer.ts
/**
 * OPA Binary Installer for Cupcake TypeScript Bindings
 *
 * This module handles automatic download and verification of the OPA binary
 * required for policy compilation.
 *
 * Version: OPA v1.7.1
 */

import * as crypto from 'crypto';
import * as fs from 'fs';
import * as os from 'os';
import * as path from 'path';
import { promisify } from 'util';
import { execFile } from 'child_process';

const execFileAsync = promisify(execFile);

// OPA Version Configuration
const OPA_VERSION = 'v0.70.0';
const OPA_BASE_URL = `https://github.com/open-policy-agent/opa/releases/download/${OPA_VERSION}`;

interface BinaryInfo {
  binary: string;
  sha256: string;
  size_mb: number;
}

// Platform-specific binary mapping with SHA256 checksums
const OPA_BINARIES: Record<string, BinaryInfo> = {
  'darwin-x64': {
    binary: 'opa_darwin_amd64',
    sha256: '51da8fa6ce4ac9b963d4babbd78714e98880b20e74f30a3f45a96334e12830bd',
    size_mb: 67.3,
  },
  'darwin-arm64': {
    binary: 'opa_darwin_arm64_static',
    sha256: 'fe2a14b6ba7f587caeb62ef93ef62d1e713776a6e470f4e87326468a8ecfbfbd',
    size_mb: 43.8,
  },
  'linux-x64': {
    binary: 'opa_linux_amd64',
    sha256: '7426bf5504049d7444f9ee9a1d47a64261842f38f5308903ef6b76ba90250b5a',
    size_mb: 67.1,
  },
  'linux-arm64': {
    binary: 'opa_linux_arm64_static',
    sha256: 'a81af8cd767f1870e9e23b8ed0ad8f40b24e5c0a64c5768c75d5c292aaa81e54',
    size_mb: 43.2,
  },
  'win32-x64': {
    binary: 'opa_windows_amd64.exe',
    sha256: '205f87d0fd1e2673c3a6f9caf9d9655290e478a93eeb3ef9f211acdaa214a9ca',
    size_mb: 98.7,
  },
};

/**
 * Get the directory for caching OPA binaries
 */
function getCacheDir(): string {
  let cacheBase: string;

  if (process.platform === 'win32') {
    cacheBase = process.env.LOCALAPPDATA || path.join(os.homedir(), 'AppData', 'Local');
  } else {
    cacheBase = process.env.XDG_CACHE_HOME || path.join(os.homedir(), '.cache');
  }

  const cacheDir = path.join(cacheBase, 'cupcake', 'bin');
  fs.mkdirSync(cacheDir, { recursive: true });

  return cacheDir;
}

/**
 * Get platform identifier for OPA binary lookup
 */
function getPlatformKey(): string {
  const platform = process.platform;
  const arch = process.arch;

  // Normalize platform and arch
  const platformMap: Record<string, string> = {
    darwin: 'darwin',
    linux: 'linux',
    win32: 'win32',
  };

  const archMap: Record<string, string> = {
    x64: 'x64',
    arm64: 'arm64',
  };

  const normalizedPlatform = platformMap[platform];
  const normalizedArch = archMap[arch];

  if (!normalizedPlatform || !normalizedArch) {
    throw new Error(`Unsupported platform: ${platform} ${arch}`);
  }

  return `${normalizedPlatform}-${normalizedArch}`;
}

/**
 * Verify the SHA256 checksum of a file
 */
async function verifyChecksum(filePath: string, expectedSha256: string): Promise<boolean> {
  return new Promise((resolve, reject) => {
    const hash = crypto.createHash('sha256');
    const stream = fs.createReadStream(filePath);

    stream.on('data', (data) => hash.update(data));
    stream.on('end', () => {
      const actual = hash.digest('hex');
      resolve(actual === expectedSha256);
    });
    stream.on('error', reject);
  });
}

/**
 * Download a file with progress indication
 */
async function downloadWithProgress(url: string, destPath: string, expectedSizeMb: number): Promise<void> {
  console.log(`Downloading OPA ${OPA_VERSION} (${expectedSizeMb.toFixed(1)} MB)...`);

  const fetch = require('node-fetch');
  const response = await fetch(url);

  if (!response.ok) {
    throw new Error(`Failed to download: ${response.statusText}`);
  }

  const totalSize = parseInt(response.headers.get('content-length') || '0', 10);
  let downloaded = 0;

  const fileStream = fs.createWriteStream(destPath);

  return new Promise((resolve, reject) => {
    response.body!.on('data', (chunk: Buffer) => {
      downloaded += chunk.length;
      fileStream.write(chunk);

      if (totalSize > 0) {
        const percent = (downloaded / totalSize) * 100;
        process.stdout.write(`Progress: ${percent.toFixed(1)}%\r`);
      }
    });

    response.body!.on('end', () => {
      fileStream.end();
      console.log('\nDownload complete!');
      resolve();
    });

    response.body!.on('error', (err: Error) => {
      fileStream.close();
      reject(err);
    });
  });
}

/**
 * Make a file executable on Unix-like systems
 */
async function makeExecutable(filePath: string): Promise<void> {
  if (process.platform !== 'win32') {
    await fs.promises.chmod(filePath, 0o755);
  }
}

/**
 * Download and verify the OPA binary for the current platform
 *
 * @param force - Force re-download even if binary exists
 * @returns Path to the OPA executable
 */
export async function downloadOpa(force = false): Promise<string> {
  // Get platform info
  const platformKey = getPlatformKey();

  if (!(platformKey in OPA_BINARIES)) {
    throw new Error(
      `Unsupported platform: ${platformKey}\n` + `Supported platforms: ${Object.keys(OPA_BINARIES).join(', ')}`,
    );
  }

  const binaryInfo = OPA_BINARIES[platformKey];
  const { binary: binaryName, sha256: expectedSha256, size_mb: sizeMb } = binaryInfo;

  // Determine local path
  const cacheDir = getCacheDir();
  let localName = `opa-${OPA_VERSION}`;
  if (process.platform === 'win32') {
    localName += '.exe';
  }
  const localPath = path.join(cacheDir, localName);

  // Check if already downloaded and valid
  if (fs.existsSync(localPath) && !force) {
    console.log(`Verifying existing OPA binary at ${localPath}...`);
    if (await verifyChecksum(localPath, expectedSha256)) {
      console.log('Checksum verified successfully!');
      return localPath;
    } else {
      console.log('Checksum verification failed, re-downloading...');
      await fs.promises.unlink(localPath);
    }
  }

  // Download the binary
  const url = `${OPA_BASE_URL}/${binaryName}`;
  const tempPath = localPath + '.tmp';

  try {
    await downloadWithProgress(url, tempPath, sizeMb);

    // Verify checksum
    console.log('Verifying checksum...');
    if (!(await verifyChecksum(tempPath, expectedSha256))) {
      throw new Error(
        `Checksum verification failed for ${binaryName}\n` + `This could indicate a corrupted download or security issue.`,
      );
    }

    // Move to final location
    await fs.promises.rename(tempPath, localPath);
    await makeExecutable(localPath);

    console.log(`OPA ${OPA_VERSION} installed successfully at ${localPath}`);
    return localPath;
  } catch (error) {
    // Clean up on failure
    if (fs.existsSync(tempPath)) {
      await fs.promises.unlink(tempPath);
    }
    throw new Error(`Failed to download OPA: ${error}`);
  }
}

/**
 * Find OPA binary in order of preference:
 * 1. Cached download
 * 2. System PATH
 *
 * @returns Path to OPA or null if not found
 */
async function findOpa(): Promise<string | null> {
  // Check cache
  const cacheDir = getCacheDir();
  let localName = `opa-${OPA_VERSION}`;
  if (process.platform === 'win32') {
    localName += '.exe';
  }
  const cachedPath = path.join(cacheDir, localName);

  if (fs.existsSync(cachedPath)) {
    return cachedPath;
  }

  // Check system PATH
  const opaCmd = process.platform === 'win32' ? 'opa.exe' : 'opa';

  try {
    const { stdout } = await execFileAsync(process.platform === 'win32' ? 'where' : 'which', [opaCmd]);
    const opaPath = stdout.trim().split('\n')[0];

    if (opaPath && fs.existsSync(opaPath)) {
      // For compatibility, accept any OPA version that can be executed
      // The version check is primarily for ensuring the downloaded binary is correct
      try {
        const { stdout: versionOutput } = await execFileAsync(opaPath, ['version'], { timeout: 5000 });
        // If we can get version output, OPA is functional
        if (versionOutput && versionOutput.includes('Version:')) {
          return opaPath;
        }
      } catch {
        // Version check failed, continue
      }
    }
  } catch {
    // which/where failed, OPA not in PATH
  }

  return null;
}

/**
 * Ensure OPA is installed and return its path
 *
 * This is the main entry point for the installer.
 * It will use an existing OPA if found, or download it if needed.
 *
 * @returns Path to the OPA executable
 */
export async function ensureOpaInstalled(): Promise<string> {
  // Try to find existing OPA
  const opaPath = await findOpa();
  if (opaPath) {
    return opaPath;
  }

  // Download OPA
  return downloadOpa();
}

/**
 * Get the OPA command for subprocess calls
 *
 * @returns String path to OPA executable
 */
export async function getOpaCommand(): Promise<string> {
  return ensureOpaInstalled();
}
